src/modules/roc_sdp/parser.rl

```diff
@@ -0,0 +1,283 @@
+    const char* p = str;
+    const char *pe = str + strlen(str);
+
+    const char *eof = pe;
+    int cs = 0;
+
+    // for start_token
+    const char* start_p = NULL;
+
+    // parse result
+    bool success = false;
+
+    const char* start_p_origin_username = NULL;
+    const char* end_p_origin_sess_id = NULL;
+    const char* start_p_origin_nettype = NULL;
+    const char* end_p_origin_addr = NULL;
+
+    // Address type of the current address being parsed.
+    address::AddrFamily cur_addrtype = address::Family_Unknown;
+
+    %%{
+
+        action start_token {
+            start_p = p;
+        }
+
+        action set_origin_unicast_address {
+            if(!result.set_origin_unicast_address(cur_addrtype, start_p, p - start_p)) {
+                roc_log(LogError, "sdp: parse origin: invalid unicast address");
+                result.clear();
+                return false;
+            }
+        }
+
+        action set_guid {
+            if(!result.set_guid(start_p_origin_username, 
+                                end_p_origin_sess_id,
+                                start_p_origin_nettype,
+                                end_p_origin_addr)) {
+                roc_log(LogError, "sdp: parse guid: invalid origin field");
+                result.clear();
+                return false;
+            }
+        }
+
+        action set_session_connection_data {
+            if(!result.set_session_connection_data(
+                    cur_addrtype,
+                    start_p, 
+                    p - start_p)) {
+                        
+                roc_log(LogError, "sdp: parse session connection data: invalid connection address");
+                result.clear();
+                return false;
+            }
+        }
+
+        action create_media {
+            if(!result.create_media_description()) {
+                roc_log(LogError, "sdp: parse media: impossible to add a new media description");
+                result.clear();
+                return false;
+            }
+        }
+
+        action set_media_port {
+
+            char* end_p = NULL;
+            long port = strtol(start_p, &end_p, 10);
+
+            if (port == LONG_MAX || port == LONG_MIN || end_p != p) {
+                roc_log(LogError, "sdp: parse media: invalid port");
+                return false;
+            }
+
+            if (!result.last_media_description()->set_port((int)port)) {
+                roc_log(LogError, "sdp: parse media: invalid port");
+                return false;
+            }
+        }
+
+        action set_media_nb_ports {
+
+            char* end_p = NULL;
+            long nb_ports = strtol(start_p, &end_p, 10);
+
+            if (nb_ports == LONG_MAX || nb_ports == LONG_MIN || end_p != p) {
+                roc_log(LogError, "sdp: parse media: invalid number of ports");
+                return false;
+            }
+
+            if (!result.last_media_description()->set_nb_ports((int)nb_ports)) {
```

Comment:
We always use `last_media_description()` without checking whether it returned NULL. Is it guaranteed by our grammar that we can call it only after at least one `create_media_description()` call? I guess yes, just to be sure.

Reply:
Yes it's guaranteed by our grammar: add_media is always called first, and if the media could not be created, we return false and stop the parsing.

Reply:
Thanks, great.